### PR TITLE
Better output for combined errors

### PR DESCRIPTION
**Pull request by erikh**
Signed-off-by: Erik Hollensbe github@hollensbe.org

### CODE REVIEW

**Comment by erikh** [comment]
Output looks like this:

```
an error
        github.com/contiv/errored.getError [errored_test.go 32]
        github.com/contiv/errored.TestErrorStackTrace [errored_test.go 37]
        testing.tRunner [testing.go 473]
        runtime.goexit [asm_amd64.s 1998]
error 2
        github.com/contiv/errored.TestErrorStackTrace [errored_test.go 76]
        testing.tRunner [testing.go 473]
        runtime.goexit [asm_amd64.s 1998]
```

Before it would combine the stack traces and print the primary error only.

**Comment by erikh** [comment]
I just reversed the order so that the primary error is printed first, cascading backwards through the combined errors.

File: errored.go [hunk 1]
```diff
@@ -152,11 +152,13 @@ func (e *Error) Error() string {
...
 	if e.trace || AlwaysTrace {
-		ret := desc + "\n"
-
-		for _, stack := range e.stack {
-			for _, line := range stack {
-				ret += fmt.Sprintf("\t%s [%s %d]\n", line.fun, line.file, line.line)
+		var ret string
+		for i := 0; i < len(e.errors); i++ {
+			ret += fmt.Sprintf("%s\n", e.errors[i].Error())
+			if len(e.stack) > i {
```
Comment (yuva29) [inline errored.go]: Isn't there a stack trace for all errors?

Comment (mapuri) [inline errored.go]: @yuva29 , I think we only keep stack for errors of type `errored.Error`

@erikh 
I might be missing something here, but I think if we were to combine a mix of errored and non-errored errors then the stack trace association might not come out correct in some cases:

``` go
func func1 () error {
    return errored.New("foobar")
}

func func1Cleanup() error {
    return fmt.Errorf("bar")
}

func func2 () error {
   if e1 = func1(); e1 != nil {
      if e2 = func1Cleanup(); e2 != nil  {
            return errored.New("foo").Combine(e2).Combine(e1)  // the stack for this resulting error will mismatch the stacked errors
      }
      return errored.New("foo").Combine(e1)
   }
}
...
...
```

Comment (erikh) [inline errored.go]: `Combine()` accepts error but it converts to `*Error` via encapsulation, so all is good :) We even get a stack trace at `Combine` time.

Comment (mapuri) [inline errored.go]: @erikh 

> We even get a stack trace at Combine time.

I am not sure if this is true in the current code for **non-Errored** errors from line: https://github.com/contiv/errored/pull/11/files/180ad0e9df6ff1fe3a629add66e0f69d9938a66c#diff-57f00f2899a684aeeac51afa14c217e2L140

I quickly tried a dummy test in this commit (https://github.com/contiv/errored/commit/b5057a2e4ccbce1866a8c8c1087fed8c1709001f) to verify my understanding that I shared above and this is what I see. Note that the stack for errored error `bar` get's incorrectly reported against non-errored error `foobar`. Also the panic is the result of a mismatch between number of combined errors and the number of combined stack traces:

```
--- FAIL: TestErrorCombined (0.00s)
        errored_test.go:138: mismatch between combined errors and combined stacks
        errored_test.go:141: foo: [{errored_test.go 136 github.com/contiv/errored.TestErrorCombined} {testing.go 473 testing.tRunner} {asm_amd64.s 1998 runtime.goexit}]
        errored_test.go:141: foobar: [{errored_test.go 136 github.com/contiv/errored.TestErrorCombined} {testing.go 473 testing.tRunner} {asm_amd64.s 1998 runtime.goexit}]
panic: runtime error: index out of range [recovered]
        panic: runtime error: index out of range
```

Comment (erikh) [inline errored.go]: It should be doing this and I'm not sure why it isn't. I'll look into it.

I'll paste this test into my branch unless you want the commit; then you can PR me. :D

Comment (mapuri) [inline errored.go]: go ahead @erikh , feel free to use this test.

**Comment by erikh** [comment]
@mapuri I have adjusted the output and the code to accommodate the panic and some cleanups along the way.

output looks like this now:

```
=== RUN   TestErrorCombined
foo
        github.com/contiv/errored.TestErrorCombined [errored_test.go 136]
        testing.tRunner [testing.go 473]
        runtime.goexit [asm_amd64.s 1998]
foobar
        github.com/contiv/errored.TestErrorCombined [errored_test.go 136]
        testing.tRunner [testing.go 473]
        runtime.goexit [asm_amd64.s 1998]
bar
        github.com/contiv/errored.TestErrorCombined [errored_test.go 136]
        testing.tRunner [testing.go 473]
        runtime.goexit [asm_amd64.s 1998]
```

File: errored.go (hunk 1 above)
Comment (mapuri) [inline errored.go]: @erikh i think we don't need the `if len(e.stack) > i` check here with latest changes ?

**Comment by mapuri** [comment]
thanks @erikh 

LGTM, other than one minor comment.

**Comment by erikh** [comment]
I'll look at that tonight and see if it's still necessary.

**Comment by erikh** [comment]
let me know if this works for you @mapuri. I've made the change and our test suite is still passing so 👍

**Comment by mapuri** [comment]
LGTM. This should be ok to merge!

**Pull request merged**